plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -223,31 +218,164 @@ func (g *genericScheduler) Preempt(pod *v1.Pod, nodeLister algorithm.NodeLister,
+			"preemption failed: the target node %s has been deleted from scheduler cache",
+			candidateNodeName)
+	}
+}
+
+// processPreemptionWithExtenders processes preemption with extenders
+func (g *genericScheduler) processPreemptionWithExtenders(
+	pod *v1.Pod,
+	nodeToVictims map[string]*schedulerapi.Victims,
+	pdbs []*policy.PodDisruptionBudget,
+) (string, error) {
+	// The final candidate node for preemption.
+	var candidateNodeName string
+
+	// Scheduler extenders that do not support preemption
+	nonPreemptingExtenders := []algorithm.SchedulerExtender{}
+
+	if len(nodeToVictims) > 0 {
+		for _, extender := range g.extenders {
+			if extender.SupportsPreemption() {
+				if err := g.checkNodesWithPreemptingExtenders(pod, nodeToVictims, pdbs, extender); err != nil {
+					return "", err
+				}
+				// If node list has all removed, no preemption will happen, skip other extenders.
+				if len(nodeToVictims) == 0 {
+					break
+				}
+			} else {
+				nonPreemptingExtenders = append(nonPreemptingExtenders, extender)
+			}
+		}
+	}
+
+	if len(nonPreemptingExtenders) == 0 {
+		// If no nonPreemptingExtenders left, just pick one best node
+		// from nodeToVictims as preemption candidate.
+		candidateNodeName = pickOneNodeForPreemption(nodeToVictims)
+	} else {
+		// If preemption is unsupported for some extenders,
+		// fall back to check if candidate node is allowed by those extenders.
+		candidateNodeName = g.pickOneNodeWithNonPreemptingExtenders(pod, nodeToVictims, nonPreemptingExtenders)
+	}
+
+	return candidateNodeName, nil
+}
+
+// checkNodesWithPreemptingExtenders check nodeToVictims by extenders which support ProcessPreemption(),
+// and modify nodeToVictims based on the check.
+func (g *genericScheduler) checkNodesWithPreemptingExtenders(
+	pod *v1.Pod,
+	nodeToVictims map[string]*schedulerapi.Victims,
+	pdbs []*policy.PodDisruptionBudget,
+	extender algorithm.SchedulerExtender) error {
+	// Pass nodeToVictims to extender to do extra preemption logics.
+	// We can expect it returns a nodeToVictimsAfterExtender struct with:
+	//   1. Subset of the given nodes after Filter.
+	//   2. Additional victims on the nodes after preemption.
+	if nodeToVictimsAfterExtender, err := extender.ProcessPreemption(pod,
+		nodeToVictims, g.cachedNodeInfoMap, pdbs); err != nil {
+		return err
+	} else {
+		// Write back result after extender to nodeToVictims.
+		for nodeName := range nodeToVictims {
+			// If node exists in nodeToVictims but does not exist in nodeToVictimsAfterExtender, delete it.
+			if victimAfterExtender, ok := nodeToVictimsAfterExtender[nodeName]; !ok {
+				delete(nodeToVictims, nodeName)
+			} else {
+				nodeToVictims[nodeName] = victimAfterExtender
+			}
+		}
+	}
+
+	return nil
+}
+
+// pickOneNodeWithNonPreemptingExtenders checks nodeToVictims by extenders which do not support ProcessPreemption(),
+// and returns the candidate node after checking.
+func (g *genericScheduler) pickOneNodeWithNonPreemptingExtenders(
```

Comment:
I think `pickOneNodePassingNonPreemptingExtenders` would be a better name for the function.